sys/subr_copy.c

```diff
@@ -0,0 +1,71 @@
+#include <uio.h>
+#include <systm.h>
+#include <stdc.h>
+#include <vm_map.h>
+
+static int copyin_vmspace(vm_map_t *vm, const void *restrict udaddr,
+                          void *restrict kaddr, size_t len) {
+  if (vm == get_kernel_vm_map()) {
+    memcpy(kaddr, udaddr, len);
+    return 0;
+  } else if (vm == get_user_vm_map()) {
+    return copyin(udaddr, kaddr, len);
+  } else {
+    /* copyin on non-active vm maps is not supported */
+    assert(false);
+    __builtin_unreachable();
+  }
+}
+static int copyout_vmspace(vm_map_t *vm, const void *restrict kaddr,
+                           void *restrict udaddr, size_t len) {
+  if (vm == get_kernel_vm_map()) {
+    memcpy(udaddr, kaddr, len);
+    return 0;
+  } else if (vm == get_user_vm_map()) {
+    return copyout(kaddr, udaddr, len);
+  } else {
+    /* copyout on non-active vm maps is not supported */
+    assert(false);
+    __builtin_unreachable();
+  }
+}
+
+/* Heavily inspired by NetBSD's uiomove */
+/* This function modifies uio to reflect on the progress. */
+int uiomove(void *buf, size_t n, uio_t *uio) {
+  char *cbuf = buf + uio->uio_offset;
+  assert(uio->uio_op == UIO_READ || uio->uio_op == UIO_WRITE);
+  int error = 0;
+  while (n > 0 && uio->uio_resid > 0) {
+    /* Take the first io vector */
+    iovec_t *iov = uio->uio_iov;
+    size_t cnt = iov->iov_len;
+    if (cnt == 0) {
+      /* If no data left to move in this vector, proceed to the next io vector,
+         or finish moving data if this was the last vector. */
+      if (uio->uio_iovcnt == 0)
+        break;
+      uio->uio_iov++;
```

Comment:
In such case, the caller must have another pointer to the allocation. In case `uiomove` fails, the passed `uio` shall represent I/O operations that were not performed.